src/DotNetty.Transport/Channels/ChannelOutboundBuffer.cs

```diff
@@ -344,11 +306,104 @@ public void RemoveBytes(long writtenBytes)
+        // See https://github.com/netty/netty/issues/3837
+        void ClearNioBuffers() => NioBuffers.Value.Clear();
+        
+        ///
+        ///Returns an array of direct NIO buffers if the currently pending messages are made of {@link ByteBuf} only.
+        ///{@link #nioBufferCount()} and {@link #nioBufferSize()} will return the number of NIO buffers in the returned
+        ///array and the total number of readable bytes of the NIO buffers respectively.
+        ///<p>
+        ///Note that the returned array is reused and thus should not escape
+        ///{@link AbstractChannel#doWrite(ChannelOutboundBuffer)}.
+        ///Refer to {@link NioSocketChannel#doWrite(ChannelOutboundBuffer)} for an example.
+        ///</p>
+        ///
+        public List<ArraySegment<byte>> GetNioBuffers()
+        {
+            long nioBufferSize = 0;
+            int nioBufferCount = 0;
+            InternalThreadLocalMap threadLocalMap = InternalThreadLocalMap.Get();
+            List<ArraySegment<byte>> nioBuffers = NioBuffers.Get(threadLocalMap);
+            Entry entry = this.flushedEntry;
+            while (this.IsFlushedEntry(entry) && entry.Message is IByteBuffer) {
+                if (!entry.Cancelled)
+                {
+                    var buf = (IByteBuffer)entry.Message;
+                    int readerIndex = buf.ReaderIndex;
+                    int readableBytes = buf.WriterIndex - readerIndex;
+
+                    if (readableBytes > 0)
+                    {
+                        if (int.MaxValue - readableBytes < nioBufferSize)
+                        {
+                            // If the nioBufferSize + readableBytes will overflow an Integer we stop populate the
+                            // ByteBuffer array. This is done as bsd/osx don't allow to write more bytes then
+                            // Integer.MAX_VALUE with one writev(...) call and so will return 'EINVAL', which will
+                            // raise an IOException. On Linux it may work depending on the
+                            // architecture and kernel but to be safe we also enforce the limit here.
+                            // This said writing more the Integer.MAX_VALUE is not a good idea anyway.
+                            //
+                            // See also:
+                            // - https://www.freebsd.org/cgi/man.cgi?query=write&sektion=2
+                            // - http://linux.die.net/man/2/writev
+                            break;
+                        }
+                        nioBufferSize += readableBytes;
+                        int count = entry.Count;
+                        if (count == -1)
+                        {
+                            //noinspection ConstantValueVariableUse
+                            entry.Count = count = buf.IoBufferCount;
+                        }
+                        if (count == 1)
```

Comment:
I'm using list which does that internally. And I need to use list because Socket.Send accepts IList<ArraySegment<byte>>> meaning that there's no way to supply different length along with array.